eeps/eep-0049.md

```diff
@@ -784,6 +773,154 @@ could result in either a success or error. The `~>>` operator matches
+success while it was actually a complete failure to match and handle the
+information given.  This is made even more ambiguous when data has the
+right shape and type, but a set of bound variables ultimately define
+whether the match succeeds or fails (in the case of a UDP socket,
+returning values that comes from the wrong peer, for example).
+
+In worst cases, It could let raw unformatted data exit a conditional
+pipeline with no way to detect it after the fact, particularly if later
+functions in `begin ... end` apply transformations to text, such as
+anonymizing or sanitizing data. This could be pretty unsafe
+and near impossible to debug well.
+
+Think for example of:
+
+    -spec fetch() -> {ok, iodata()} | {error, _}.
+    fetch() ->
+        begin
+            {ok, B = <<_/binary>>} <- f(),
+            true <- validate(B),
+            {ok, sanitize(B)}
+        end.
+
+If the value returned from `f()` turns out to be a list (say it’s a
+misconfigured socket using `list` instead of `binary` as an option), the
+expression will return early, the `fetch()` function will still return
+`{ok, iodata()}` but you couldn’t know as a caller whether it is the
+transformed data or non-matching content. It would not be obvious to
+most developers either that this could represent a major security risk
+by allowing unexpected data to be seen as clean data.
+
+This specific type of error is in fact possible in Elixir, but no such
+warning appears to have been circulating within its community so far.
+The issue is to be handled with an `else` block which this proposal
+reuses to clamp down on unexpected values:
+
+    -spec fetch() -> {ok, iodata()} | {error, _}.
+    fetch() ->
+        begin
+            {ok, B = <<_/binary>>} <- f(),
+            true <- validate(B),
+            {ok, sanitize(B)}
+        else
+            false -> {error, invalid_data};
+            {error, R} -> {error, R}
+        end.
+
+Here misconfigured sockets won’t result in unchecked data passing trough
+your app; any invalid use case is captured, and if the value for `B` turns
+out to be a list, an `else_clause` error is raised with the bad value.
+
+Unless the clause is mandatory (it is not in Elixir and we do not plan it here
+either for compatibility reasons with existing `begin ... end` expressions),
+this level of additional matching is purely optional; the developer has no
+obvious incentive to go and handle these errors, and if they do, the exception
+raised will be through a missing clause in the `else` section, which will
+obscure its origin and line number.
+
+We will therefore have to rely on education and documentation (along with
+type analysis) to prevent such issues from arising in the future.
+
+These problems would *not* exist with normalized error and return values as those used in statically-typed languages, but since we do not intend to normalize values, the `else` blog is a necessary workaround.
```

Comment:
```suggestion
These problems would *not* exist with normalized error and return values as those used in statically-typed languages, but since we do not intend to normalize values, the `else` block is a necessary workaround.
```